File: JavaScript/ExampleTests/common/NewPostString.js
import * as tokens from "./tokens.js";

export class NewPostString {
  static DNC_STRING = "?DNC?";

  constructor(title = "", body = "", userId = "") {
    this.title = title;
    this.body = body;
    this.userId = userId;
  }

  static fromList(values) {
    const v = Array.from(values);
    return new NewPostString(
      v[0] ?? "",
      v[1] ?? "",
      v[2] ?? ""
    );
  }

  /** Builds from the text form, e.g. Money as "25 USD". */
  static fromText(text) {
    const parts = tokens.require_(text, 3, "NewPost");
    return new NewPostString(
      parts[0],
      parts[1],
      parts[2]
    );
  }

  toString() {
    return tokens.token(this.title) + " " + tokens.token(this.body) + " " + tokens.token(this.userId);
  }

  equals(other) {
    if (!(other instanceof NewPostString)) return false;
    return (this.title === NewPostString.DNC_STRING || other.title === NewPostString.DNC_STRING || this.title === other.title)
      && (this.body === NewPostString.DNC_STRING || other.body === NewPostString.DNC_STRING || this.body === other.body)
      && (this.userId === NewPostString.DNC_STRING || other.userId === NewPostString.DNC_STRING || this.userId === other.userId);
  }

  static equalLists(a, b) {
    if (a.length !== b.length) return false;
    return a.every((row, i) => row.equals(b[i]));
  }
}

File: Java/SpectableJavaExamplesTest/src/test/java/spectable/common/NewPostString.java
package spectable.common;

import java.util.Objects;
import production.*;
import records.*;
import calculator.*;

public class NewPostString {
    private static final String DNCString = "?DNC?";

    public String title;
    public String body;
    public String userId;

    public NewPostString(String title, String body, String userId) {
        this.title = title;
        this.body = body;
        this.userId = userId;
    }

    /** Builds from the text form, e.g. Money as "25 USD". */
    public static NewPostString fromText(String text) {
        java.util.List<String> parts = Tokens.require(text, 3, "NewPost");
        return new NewPostString(parts.get(0), parts.get(1), parts.get(2));
    }

    public NewPostString(String text) {
        NewPostString parsed = fromText(text);
        this.title = parsed.title;
        this.body = parsed.body;
        this.userId = parsed.userId;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof NewPostString)) return false;
        NewPostString that = (NewPostString) o;
        return (DNCString.equals(title) || DNCString.equals(that.title) || Objects.equals(title, that.title))
            && (DNCString.equals(body) || DNCString.equals(that.body) || Objects.equals(body, that.body))
            && (DNCString.equals(userId) || DNCString.equals(that.userId) || Objects.equals(userId, that.userId));
    }

    @Override
    public int hashCode() {
        return Objects.hash(title, body, userId);
    }

    @Override
    public String toString() {
        return Tokens.token(title) + " " + Tokens.token(body) + " " + Tokens.token(userId);
    }
}
